src/klein/_app.py

```diff
@@ -11,32 +11,53 @@
 try:
     from inspect import iscoroutine
 except ImportError:
-    def iscoroutine(*args, **kwargs):  # type: ignore
+    def iscoroutine(obj):  # type: ignore[misc]
+        # type: (object) -> bool
         return False
+from typing import (
+    Any, Callable, Dict, IO, List, Mapping, Optional, Text, Union, cast
+)
 from weakref import ref
 
 from twisted.internet import endpoints, reactor
+from twisted.internet.defer import Deferred
 from twisted.python import log
 from twisted.python.components import registerAdapter
+from twisted.python.failure import Failure
+from twisted.web.iweb import IRenderable, IRequest
+from twisted.web.resource import IResource
 from twisted.web.server import Request, Site
 
 try:
     from twisted.internet.defer import ensureDeferred
 except ImportError:
-    def ensureDeferred(*args, **kwagrs):
+    def ensureDeferred(coro):
+        # type: (Awaitable) -> Deferred
         raise NotImplementedError("Coroutines support requires Twisted>=16.6")
 
-from werkzeug.routing import Map, Rule, Submount
+from werkzeug.routing import Map, MapAdapter, Rule, Submount
 
 from zope.interface import implementer
 
 from ._decorators import modified, named
 from ._interfaces import IKleinRequest
 from ._resource import KleinResource
+from ._typing import Awaitable, KwArg, VarArg
 
 
+KleinSynchronousRenderable = Union[str, bytes, IResource, IRenderable]
+KleinRenderable = Union[
+    KleinSynchronousRenderable, Awaitable[KleinSynchronousRenderable]
+]
+KleinRoute = Callable[
+    [Any, IRequest, VarArg(Any), KwArg(Any)],
+    KleinRenderable
+]
```

Comment:
Initially called this `KleinEndpoint`, but I was thinking of using `@endpoint` as the `txrequest`-style decorator; this is for `@route`-decorated callables.